### PR TITLE
setup.py for building C++ backend

### PR DESCRIPTION
**Pull request by 1e-to**
This PR must resolve the problem from this issue #145.

### CODE REVIEW

File: conda-recipe/bld.bat [hunk 1]
```diff
@@ -1,41 +1,41 @@
...
+@REM ninja -n
+@REM ninja install
+@REM IF %ERRORLEVEL% NEQ 0 exit /b 1
+
+@REM cd ..
+@REM xcopy install\lib\*.lib dpctl /E /Y
+@REM xcopy install\bin\*.dll dpctl /E /Y
+
+@REM mkdir dpctl\include
+@REM xcopy backends\include dpctl\include /E /Y
```
Comment (PokhodenkoSA) [inline conda-recipe/bld.bat:38]: Remove old code.

File: conda-recipe/build.sh [hunk 2]
```diff
@@ -1,44 +1,44 @@
...
+#     -DNUMPY_INCLUDE_DIR=${NUMPY_INC}                        \
+#     ../backends
+
+# make -j 4 && make install
+
+# popd
+# cp install/lib/*.so dpctl/
+
+# mkdir -p dpctl/include
+# cp -r backends/include/* dpctl/include
```
Comment (PokhodenkoSA) [inline conda-recipe/build.sh:41]: Remove old code.

File: setup.py [hunk 3]
```diff
@@ -99,7 +120,15 @@ def get_suppressed_warning_flags():
         return []
 
 
+def build_backend():
+    if IS_LIN:
+        subprocess.check_call(["/bin/bash", "-c", "scripts/build_backend.sh"])
+    elif IS_WIN:
+        subprocess.check_call(["cmd.exe", "/c", "scripts\\build_backend.bat"])
```
Comment (PokhodenkoSA) [inline setup.py:127]: If build scripts are Python then no need to have 2 versions for Linux and Windows.

File: setup.py [hunk 4]
```diff
@@ -99,7 +120,15 @@ def get_suppressed_warning_flags():
...
+
+
 def extensions():
+    build_backend()
```
Comment (PokhodenkoSA) [inline setup.py:131]: It is not a good place to run build scripts for backend.

**Review by PokhodenkoSA — CHANGES_REQUESTED** [review]

**Comment by oleksandr-pavlyk** [comment]
I thought this PR would build on Python's interface to CMake:

https://github.com/pytorch/pytorch/blob/master/tools/build_pytorch_libs.py#L51

Kind of like 

https://github.com/pytorch/pytorch/blob/master/tools/setup_helpers/cmake.py#L94

Maybe a little too heavy-handed for our purposes.

File: setup.py [hunk 5]
```diff
@@ -161,10 +184,29 @@ def extensions():
...
+class install(_install):
+    def run(self):
+        build_backend()
+        return super().run()
+
+
+class develop(_develop):
+    def run(self):
+        build_backend()
+        return super().run()
```
Comment (PokhodenkoSA) [inline setup.py:196]:
```suggestion
class install(setuptools.command.install.install):
    def run(self):
        build_backend()
        return super().run()


class develop(setuptools.command.develop.develop):
    def run(self):
        build_backend()
        return super().run()
```

File: setup.py [hunk 6]
```diff
@@ -25,8 +25,11 @@
 import os.path
 import sys
 import versioneer
+import subprocess
 
 from setuptools import setup, Extension, find_packages
+from setuptools.command.install import install as _install
+from setuptools.command.develop import develop as _develop
```
Comment (PokhodenkoSA) [inline setup.py:32]: If it is used in one place then remove it. See comment below.

File: setup.py [hunk 7]
```diff
@@ -99,6 +116,12 @@ def get_suppressed_warning_flags():
         return []
 
 
+def build_backend():
+    dpctl_dir = os.getcwd()
+    build_script = os.path.join(dpctl_dir, "scripts", "build_backend.py")
```
Comment (PokhodenkoSA) [inline setup.py:121]:
```suggestion
    build_script = os.path.join(os.getcwd(), "scripts", "build_backend.py")
```

File: scripts/build_backend.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    subprocess.check_call(
+        [
+            "cmake",
+            "-DCMAKE_BUILD_TYPE=Release",
+            "-DCMAKE_INSTALL_PREFIX=" + INSTALL_PREFIX,
+            "-DCMAKE_PREFIX_PATH=" + INSTALL_PREFIX,
+            "-DDPCPP_ROOT=" + DPCPP_ROOT,
+            "-DCMAKE_C_COMPILER:PATH=" + os.path.join(DPCPP_ROOT, "bin", "clang"),
+            "-DCMAKE_CXX_COMPILER:PATH=" + os.path.join(DPCPP_ROOT, "bin", "dpcpp"),
+            backends_dir,
```
Comment (PokhodenkoSA) [inline scripts/build_backend.py:47]: Make a list var please with all args and send the list to subprocess call. Also using `f"-DCMAKE_INSTALL_PREFIX={INSTALL_PREFIX}"` is shorter.

File: setup.py [hunk 9]
```diff
@@ -48,6 +51,20 @@
 else:
     assert False, sys.platform + " not supported"
 
+if IS_LIN:
+    DPCPP_ROOT = os.environ["ONEAPI_ROOT"] + "/compiler/latest/linux"
+    os.environ["CC"] = DPCPP_ROOT + "/bin/clang"
+    os.environ["CXX"] = DPCPP_ROOT + "/bin/dpcpp"
```
Comment (PokhodenkoSA) [inline setup.py:57]:
```suggestion
    os.environ["CXX"] = DPCPP_ROOT + "/bin/clang++"
```

File: setup.py [hunk 10]
```diff
@@ -48,6 +51,20 @@
...
+    os.environ["DPPL_SYCL_INTERFACE_LIBDIR"] = "dpctl"
+    os.environ["DPPL_SYCL_INTERFACE_INCLDIR"] = "dpctl/include"
+    os.environ["CFLAGS"] = "-fPIC"
+
+elif IS_WIN:
+    os.environ["CC"] = "clang-cl.exe"
+    os.environ["CXX"] = "dpcpp.exe"
```
Comment (PokhodenkoSA) [inline setup.py:64]: I think next step (in another PR) should be to try allow Python to build Cython modules with default compiler (not DPC++).

**Comment by PokhodenkoSA** [comment]
Tested locally on Linux: OK.
Steps:
1. `source /opt/intel/oneapi/compiler/latest/env/vars.sh`
2. `python setup.py develop`
3. `python -m unittest -v dpctl.tests`

File: setup.py (hunk 5 above)
Comment (1e-to) [reply setup.py:196]: That option don't work:
AttributeError: module 'setuptools.command' has no attribute 'install'

File: setup.py (hunk 7 above)
Comment (1e-to) [reply setup.py:121]: dpctl_dir is using for the all script

Comment (PokhodenkoSA) [reply setup.py:121]: Not in this place.

File: setup.py (hunk 5 above)
Comment (PokhodenkoSA) [reply setup.py:196]: Then `import setuptools.command.install` before.

File: scripts/build_backend.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    subprocess.check_call(cmake, stderr=subprocess.STDOUT, shell=False)
+    subprocess.check_call(["make", "-j", "4"])
+    subprocess.check_call(["make", "install"])
+
+    os.chdir(dpctl_dir)
+    for file in glob.glob(os.path.join(dpctl_dir, "install", "lib", "*.so")):
+        shutil.copy(file, os.path.join(dpctl_dir, "dpctl"))
+
+if IS_WIN:
+    cmake = [
```
Comment (PokhodenkoSA) [inline scripts/build_backend.py:57]:
```suggestion
    cmake_args = [
```

File: setup.py [hunk 12]
```diff
@@ -25,7 +25,10 @@
 import os.path
 import sys
 import versioneer
+import subprocess
 
+from setuptools.command.install import install as orig_install
+from setuptools.command.develop import develop as orig_develop
```
Comment (PokhodenkoSA) [inline setup.py:31]:
```suggestion
import setuptools.command.install as orig_install
import setuptools.command.develop as orig_develop
```

File: setup.py [hunk 13]
```diff
@@ -161,10 +183,29 @@ def extensions():
     return exts
 
 
+class install(orig_install):
```
Comment (PokhodenkoSA) [inline setup.py:186]:
```suggestion
class install(orig_install.install):
```

File: setup.py [hunk 14]
```diff
@@ -161,10 +183,29 @@ def extensions():
...
+    def run(self):
+        build_backend()
+        return super().run()
+
+
+class develop(orig_develop):
```
Comment (PokhodenkoSA) [inline setup.py:192]:
```suggestion
class develop(orig_develop.develop):
```

**Comment by PokhodenkoSA** [comment]
@1e-to, please, update README.

File: CHANGELOG.md [hunk 15]
```diff
@@ -6,6 +6,7 @@ All notable changes to this project will be documented in this file.
 - Device descriptors "max_compute_units", "max_work_item_dimensions", "max_work_item_sizes", "max_work_group_size", "max_num_sub_groups" and "aspects" for int64 atomics inside dpctl C API and inside the dpctl.SyclDevice class.
 - MemoryUSM* classes moved to `dpctl.memory` module, added support for aligned allocation, added support for `prefetch` and `mem_advise` (sychronous) methods, implemented `copy_to_host`, `copy_from_host` and `copy_from_device` methods, pickling support, and zero-copy interoperability with Python objects which implement `__sycl_usm_array_inerface__` protocol.
 - Helper scripts to generate API documentation for both C API and Python.
+- Opportunity to build and install project with setuptools.
```
Comment (PokhodenkoSA) [inline CHANGELOG.md:9]:
```suggestion
- setup.py builds C++ backend for develop and install commands.
```

**Review by PokhodenkoSA — APPROVED** [review]

File: README.md [hunk 16]
```diff
@@ -36,6 +36,20 @@ conda build --croot=C:/tmp conda-recipe
 conda install dpctl
 ```
 
+Build and Install with setuptools
+===========
```
Comment (PokhodenkoSA) [inline README.md:40]: Make count of `=` to fit the header length.

File: scripts/build_backend.py (hunk 11 above)
Comment (PokhodenkoSA) [reply scripts/build_backend.py]: Variable name should explanation better what it is.

**Review by PokhodenkoSA — APPROVED** [review]

**Review by PokhodenkoSA — APPROVED** [review]
Thanks @1e-to. LGTM. Could be merged after CI finishes.

**Pull request merged**